### PR TITLE
Test/fix scope of pattern variables declared in statements

### PR DESCRIPTION
**Pull request by AlekseyTs**
- Check for name conflicts for a local declared with [let <identifier> = …] statement.
- Create PatternVariableBinder for ReturnStatementSyntax.
- Create PatternVariableBinder for ThrowStatementSyntax.
- Create BlockBinder for a [let] statement used as an embedded statement.
- Add SemanticModel.GetDeclaredSymbol API for LetStatementSyntax.
- Test scope of pattern variable declared in [let], [return] and [throw] statements.

Related to # 8817.

@gafter, @jaredpar, @dotnet/roslyn-compiler Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CSharpExtensions.cs [hunk 1]
```diff
@@ -1220,6 +1220,15 @@ public static ISymbol GetDeclaredSymbol(this SemanticModel semanticModel, Declar
         }
 
         /// <summary>
+        /// Given a declaration pattern syntax, get the corresponding symbol.
+        /// </summary>
+        public static ISymbol GetDeclaredSymbol(this SemanticModel semanticModel, LetStatementSyntax declarationSyntax, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpExtensions.cs]: I don't think this API should be here.

Symbols are not declared "by" the let statement, any more than they are declared by an "is" expression with a pattern on the right-hand-side. Such symbols (zero, one, or many) are instead declared by the patterns appearing *inside* the let statement.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 2]
```diff
@@ -2632,6 +2632,14 @@ internal Conversion ClassifyConversionForCast(int position, ExpressionSyntax exp
         public abstract ISymbol GetDeclaredSymbol(DeclarationPatternSyntax declarationSyntax, CancellationToken cancellationToken = default(CancellationToken));
 
         /// <summary>
+        /// Given a let statement syntax, get the corresponding symbol if the statement uses [let &lt;identifier&gt; = ...] form.
+        /// </summary>
+        /// <param name="declarationSyntax">The syntax node that declares a variable.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The symbol that was declared.</returns>
+        public abstract ISymbol GetDeclaredSymbol(LetStatementSyntax declarationSyntax, CancellationToken cancellationToken = default(CancellationToken));
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Ditto.

**Comment by gafter** [comment]
LGTM, but please clarify in the doc comment for the new SemanticModel API that it only returns a symbol for the `let identifier = expression` form.

File: src/Compilers/CSharp/Portable/CSharpExtensions.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpExtensions.cs]: This is the only place where the comment is incorrect. I'll adjust it in the next change.

**Pull request merged**